### PR TITLE
BFF CRUD Settings

### LINKED ISSUE #33: BFF CRUD Settings

### PR DESCRIPTION
**Pull request by arushwolf**
Closes #33

### CODE REVIEW

**Comment by arushwolf** [comment]
This is a test comment, seeing if the Discord bot I made works.

**Review by MsMarini — APPROVED** [review]

**Review by n898c — APPROVED** [review]

File: AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs [hunk 1]
```diff
@@ -48,4 +48,62 @@ public async Task<ApiSetting> CreateSetting(CreateSettingRequest request)
...
+
+        // Wrap request object in StringContent object
+        var content = new StringContent(
+            serializedRequest,
+            System.Text.Encoding.UTF8,
+            "application/json"
+        );
+
+        // Send request
+        var response = await _httpClient.PutAsync($"{request.SettingId}/{request.Version}", content);
```
Comment (flennaja) [inline AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs:81]:
```suggestion
        var response = await _httpClient.PutAsync(string.Empty(), content);      
```

The [api update setting route](https://github.com/WSU-4110/AppSettingsManager/blob/48e5cecdcb53e26c8dffd440d22b28bdf86f42e1/AppSettingsManagerAPI/AppSettingsManagerApi/Controllers/SettingsController.cs#L59) has no added routing

File: AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs [hunk 2]
```diff
@@ -48,4 +48,62 @@ public async Task<ApiSetting> CreateSetting(CreateSettingRequest request)
...
+        var setting = JsonSerializer.Deserialize<ApiSetting>(jsonResponse, _jsonSerializerOptions);
+
+        // "??" means "if null then"
+        return setting ?? throw new HttpRequestException();
+    }
+
+    public async Task<ApiSetting> DeleteSetting([FromRoute][Required] string settingId)
+    {
+        // Send request
+        var response = await _httpClient.DeleteAsync($"{settingId}");
```
Comment (flennaja) [inline AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs:97]:
```suggestion
        var response = await _httpClient.DeleteAsync($"settingId/{settingId}");
```

File: AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs [hunk 3]
```diff
@@ -48,4 +48,62 @@ public async Task<ApiSetting> CreateSetting(CreateSettingRequest request)
...
+        var jsonResponse = await response.Content.ReadAsStringAsync();
+
+        // Deserialize json string
+        var setting = JsonSerializer.Deserialize<ApiSetting>(jsonResponse, _jsonSerializerOptions);
+
+        // "??" means "if null then"
+        return setting ?? throw new HttpRequestException();
+    }
+
+    public async Task<ApiSetting> DeleteSetting([FromRoute][Required] string settingId)
```
Comment (flennaja) [inline AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs:94]:
```suggestion
    public async Task<IEnumberable<ApiSetting>> DeleteSetting([FromRoute][Required] string settingId)
```

This method returns a list of settings
https://github.com/WSU-4110/AppSettingsManager/blob/48e5cecdcb53e26c8dffd440d22b28bdf86f42e1/AppSettingsManagerAPI/AppSettingsManagerApi/Controllers/SettingsController.cs#L69

**Review by flennaja — CHANGES_REQUESTED** [review]
Looks good overall, but make these changes and then test to see that these work

File: AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs (hunk 3 above)
Comment (arushwolf) [reply AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs:94]: So I get the change you would like for me to commit, but can you explain the second comment you made for [Line 69](https://github.com/WSU-4110/AppSettingsManager/blob/48e5cecdcb53e26c8dffd440d22b28bdf86f42e1/AppSettingsManagerAPI/AppSettingsManagerApi/Controllers/SettingsController.cs#L69).

Comment (flennaja) [reply AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs:94]: I was just pointing you to the method in the API that you're calling

Comment (arushwolf) [reply AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs:94]: Ah, ok, but  do you need me to redefine it?? Or is my definition ok for now??

File: AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs [hunk 4]
```diff
@@ -48,4 +48,62 @@ public async Task<ApiSetting> CreateSetting(CreateSettingRequest request)
...
+    {
+        // Send request
+        var response = await _httpClient.DeleteAsync($"settingId/{settingId}");
+        response.EnsureSuccessStatusCode();
+
+        // Extract response content as json string
+        var jsonResponse = await response.Content.ReadAsStringAsync();
+
+        // Deserialize json string
+        var setting = JsonSerializer.Deserialize<ApiSetting>(jsonResponse, _jsonSerializerOptions);
```
Comment (flennaja) [inline AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs:104]:
```suggestion
        var setting = JsonSerializer.Deserialize<IEnumerable<ApiSetting>>(jsonResponse, _jsonSerializerOptions);
```

Sorry didn't catch this earlier. All looks good I think except this

File: AppSettingsManagerBff/AppSettingsManagerBff/Controllers/SettingsController.cs [hunk 5]
```diff
@@ -18,7 +18,24 @@ public SettingsController(ISettingsRepository settingsRepository)
...
+    }
+
+    [HttpPut]
+    public async Task<ApiSetting> UpdateSetting([FromRoute][Required] UpdateSettingRequest request)
+    {
+        return await _settingsRepository.UpdateSetting(request);
+    }
+
+    [HttpDelete("delete/settingId/{settingId}")]
+    public async Task<ApiSetting> DeleteSetting([FromRoute][Required] string settingId)
```
Comment (flennaja) [inline AppSettingsManagerBff/AppSettingsManagerBff/Controllers/SettingsController.cs:37]:
```suggestion
    public async Task<IEnumberable<ApiSetting>> DeleteSetting([FromRoute][Required] string settingId)
```

File: AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs (hunk 3 above)
Comment (flennaja) [reply AppSettingsManagerBff/AppSettingsManagerBff.Infrastructure/ApiRepositories/HttpSettingsRepository.cs]: This is good now. See my other comments

**Comment by arushwolf** [comment]
Getting errors when trying to run AppSettingsManagerBff.csproj. Will try to meet up before class  to fix the issue, hitting the sack for the night.

**Review by flennaja — APPROVED** [review]
gotta get this moving

**Pull request merged**